### PR TITLE
Major fix for Points Distribution

### PR DESCRIPTION
**Pull request by Aritra8438**
# Description

https://github.com/user-attachments/assets/8b3cc60f-c39e-4b59-a74f-742f9c90068c

The following strategy seems to handle all the corner cases. Also, this should follow the workflow that users might want while writing the testcases.

There's a complete new button to validate the testcases to make sure that `group points` and the `cumulative casepoints` are same.

--- 

- Each ungrouped case will have by default 100 points (if `autoPoints` is true).
- Each group will have by default 100 points (if `autoPoints` is true).
- For a group with `n` cases:
  - `a` cases have points set by the user.
  - `b` cases have `autoPoints` set to true.
  - Thus, `a + b = n`.
- If the group's `autoPoints` is true:
  - The group has 100 points.
  - Each `notDefined` case (where `autoPoints` is true) will receive a score of `(1/b) * (100 - Σp(a))`.
  - If Σp(a) exceeds 100, the group's total points will be set to Σp(a).
  - In this case, all `notDefined` cases will receive 0 points.
- If the group's `autoPoints` is false:
  - Points set for the group should take precedence.
  - Users must allocate all the points in the group.
  - Σp(a) should not exceed the group's total points.
  
**Validation and Auto-Fix Features:**
- A validator will check if the point distribution is correct.
- There will be an option to automatically adjust points to satisfy the group's conditions.
- If users fail to meet the requirements:
  - The validator will maintain the ratio of case points but scale them to match the group's total points.

# Checklist:

- [x] The code follows the [coding guidelines](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) of omegaUp.
- [x] The tests were executed and all of them passed.
- [x] If you are creating a feature, the new tests were added.
- [x] If the change is large (> 200 lines), this PR was split into various Pull Requests. It's preferred to create one PR for changes in controllers + unit tests in PHPUnit,  and then another Pull Request for UI + tests in Jest, Cypress or both.

### CODE REVIEW

**Comment by Aritra8438** [comment]
@carlosabcs PTAL! Needed some tests to be added.

File: frontend/www/js/omegaup/problem/creator/modules/cases.ts [hunk 1]
```diff
@@ -55,6 +55,11 @@ export const casesStore: Module<CasesState, RootState> = {
       state.layouts = uploadedState.layouts;
       state.hide = uploadedState.hide;
     },
+    validateAndFixPoints(state) {
+      // As we are doing nothing with the state directly, es-lint complains, but this is crucial to fix the points distribution.
+      // eslint-disable-next-line
+      state = assignMissingPoints(state, true);
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/problem/creator/modules/cases.ts:61]: What about changing the state through a mutation?

File: frontend/www/js/omegaup/problem/creator/modules/cases.ts [hunk 2]
```diff
@@ -84,14 +89,19 @@ export const casesStore: Module<CasesState, RootState> = {
     addCase(state, caseRequest: CaseRequest) {
       if (caseRequest.groupID === UUID_NIL) {
         // Should create a new group with the same name
+        if (caseRequest.autoPoints === true) {
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/problem/creator/modules/cases.ts:92]:
```suggestion
        if (caseRequest.autoPoints) {
```

Comment (carlosabcs) [reply frontend/www/js/omegaup/problem/creator/modules/cases.ts:92]: ?

File: frontend/www/js/omegaup/problem/creator/modules/cases.ts [hunk 3]
```diff
@@ -185,6 +195,12 @@ export const casesStore: Module<CasesState, RootState> = {
       state = assignMissingPoints(state);
     },
     updateCase(state, [oldGroupID, updateCaseRequest]: [GroupID, CaseRequest]) {
+      if (
+        updateCaseRequest.groupID === UUID_NIL &&
+        updateCaseRequest.autoPoints === true
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/problem/creator/modules/cases.ts:200]: Same here.

File: frontend/www/js/omegaup/problem/creator/modules/cases.ts [hunk 4]
```diff
@@ -734,28 +753,95 @@ export const casesStore: Module<CasesState, RootState> = {
...
+        _group.points = casesPoints;
+      } else {
+        _group.points = 100;
+      }
+      if (_group.points === 0 && notDefinedCount != 0) {
+        _group.points = 100;
+      }
+      const remainingAssignablePoints = _group.points - casesPoints;
+      _group.cases.forEach((_case) => {
+        if (_case.autoPoints === true) {
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/problem/creator/modules/cases.ts:806]: Ok, there might be a reason for this. Is this avoidable?

Comment (Aritra8438) [reply frontend/www/js/omegaup/problem/creator/modules/cases.ts:806]: @carlosabcs, which line did you mention?

File: frontend/www/js/omegaup/problem/creator/modules/cases.ts (hunk 1 above)
Comment (Aritra8438) [reply frontend/www/js/omegaup/problem/creator/modules/cases.ts:61]: Hi @carlosabcs, `validateAndFixPoints` is itself a mutation.

**Comment by Aritra8438** [comment]
@carlosabcs, PTAL!

File: frontend/www/js/omegaup/problem/creator/modules/cases.ts (hunk 1 above)
Comment (carlosabcs) [reply frontend/www/js/omegaup/problem/creator/modules/cases.ts:61]: Ok!

**Review by carlosabcs — APPROVED** [review]
LGTM

**Comment by Aritra8438** [comment]
@heduenas PTAL!

File: frontend/templates/en.lang [hunk 5]
```diff
@@ -1119,6 +1119,11 @@ problemCreatorUngroupedCases = "Ungrouped cases"
 problemCreatorUpdateAlert = "The problem has been updated succesfully."
 problemCreatorUploadZip = "Upload"
 problemCreatorUploadZipMessage = "Upload the zip file of your problem here."
+problemCreatorValidatePoints = "Validate and fix points"
+problemCreatorValidatePointsBack = "Back"
+problemCreatorValidatePointsButton = "Validate points"
+problemCreatorValidatePointsContinue = "Continue"
+problemCreatorValidatePointsWarning = "Please continue to validate and fix the points distribution across all the groups."
```
[the remark this answers is not in the document]
Comment (tvanessa) [reply frontend/templates/en.lang:1129]: Sure!

File: frontend/templates/es.lang [hunk 6]
```diff
@@ -1122,6 +1122,11 @@ problemCreatorUngroupedCases = "Casos sin grupo"
 problemCreatorUpdateAlert = "El problema se ha actualizado correctamente."
 problemCreatorUploadZip = "Subir"
 problemCreatorUploadZipMessage = "Sube el archivo zip de tu problema aquí."
+problemCreatorValidatePoints = "Validar y corregir puntos"
```
Comment (tvanessa) [inline frontend/templates/es.lang:1125]:
```suggestion
problemCreatorValidatePoints = "Validar y corregir puntos de los grupos de casos"
```

File: frontend/templates/es.lang [hunk 7]
```diff
@@ -1122,6 +1122,11 @@ problemCreatorUngroupedCases = "Casos sin grupo"
...
+problemCreatorValidatePointsBack = "Atrás"
+problemCreatorValidatePointsButton = "Validar puntos"
```
Comment (tvanessa) [inline frontend/templates/es.lang:1127]:
```suggestion
problemCreatorValidatePointsButton = "Validar puntos de los casos"
```

File: frontend/templates/en.lang [hunk 8]
```diff
@@ -1122,6 +1122,11 @@ problemCreatorUngroupedCases = "Ungrouped cases"
...
```
[the remark this answers is not in the document]
Comment (Aritra8438) [reply frontend/templates/en.lang:1129]: These might be huge based on the user input. I thought about that, but modal or toast might be too small for these errors.
In that case what should I do?
Maybe, for each group that will be modified, we can notify users, but that might be huge too!

File: frontend/templates/en.lang [hunk 9]
```diff
@@ -1122,6 +1122,11 @@ problemCreatorUngroupedCases = "Ungrouped cases"
 problemCreatorUpdateAlert = "The problem has been updated succesfully."
 problemCreatorUploadZip = "Upload"
 problemCreatorUploadZipMessage = "Upload the zip file of your problem here."
+problemCreatorValidatePoints = "Validate and fix points"
+problemCreatorValidatePointsBack = "Back"
+problemCreatorValidatePointsButton = "Validate points"
```
[the remark this answers is not in the document]
Comment (Aritra8438) [reply frontend/templates/en.lang:1127]: image
Hi @tvanessa, we should be validating and fixing per groups, but anyways the string will be too large for a dropdown button in that case. Should I continue with that?

**Comment by Aritra8438** [comment]
@tvanessa, do you have any updates on this?

File: frontend/templates/en.lang (hunk 9 above)
[the remark this answers is not in the document]
Comment (tvanessa) [reply frontend/templates/en.lang:1127]: Ok. We can keep the shorter string

File: frontend/templates/en.lang (hunk 8 above)
[the remark this answers is not in the document]
Comment (tvanessa) [reply frontend/templates/en.lang:1129]: What about something like `The distribution of points does not follow the formula: (insert formula). Please fix`. The current string is kind of the same length.

[the remark this answers is not in the document]
Comment (Aritra8438) [reply frontend/templates/en.lang:1129]: Okay, sure, I will do that.

**Comment by Aritra8438** [comment]
@tvanessa PTAL!

**Review by tvanessa — APPROVED** [review]

**Comment by Aritra8438** [comment]
@heduenas, PTAL!

**Pull request merged**